File: src/NodeParser/ParenthesizedNodeParser.ts
import ts from "typescript";
import type { Context, NodeParser } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";

export class ParenthesizedNodeParser implements SubNodeParser {
    public constructor(protected childNodeParser: NodeParser) {}

    public supportsNode(node: ts.ParenthesizedTypeNode): boolean {
        return node.kind === ts.SyntaxKind.ParenthesizedType;
    }
    public createType(node: ts.ParenthesizedTypeNode, context: Context): BaseType {
        return this.childNodeParser.createType(node.type, context);
    }
}

File: internal/parser/parenthesized_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// ParenthesizedNodeParser handles parenthesized types by delegating to the
// inner type (src/NodeParser/ParenthesizedNodeParser.ts).
type ParenthesizedNodeParser struct {
	childNodeParser NodeParser
}

func NewParenthesizedNodeParser(childNodeParser NodeParser) *ParenthesizedNodeParser {
	return &ParenthesizedNodeParser{childNodeParser: childNodeParser}
}

func (p *ParenthesizedNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindParenthesizedType
}

func (p *ParenthesizedNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	return p.childNodeParser.CreateType(node.AsParenthesizedTypeNode().Type, ctx, nil)
}
